src/controller/CHIPDeviceController.cpp

```diff
@@ -215,46 +210,67 @@ CHIP_ERROR DeviceController::Init(NodeId localDeviceId, ControllerInitParams par
     return CHIP_NO_ERROR;
 }
 
-CHIP_ERROR DeviceController::GenerateOperationalCertificates(const ByteSpan & CSR, NodeId deviceId, MutableByteSpan & cert)
+CHIP_ERROR DeviceController::GenerateOperationalCertificates(const ByteSpan & noc, MutableByteSpan & cert)
 {
-    // This code requires about 2K RAM to generate the certificates.
+    // This code requires about 1K RAM to generate the certificates.
     // The code would run as part of commissioner applications, so RAM requirements should be fine.
     // Need to analyze if this requirement could be better managed by using static memory pools.
-    chip::Platform::ScopedMemoryBuffer<uint8_t> noc;
-    ReturnErrorCodeIf(!noc.Alloc(kMaxCHIPDERCertLength), CHIP_ERROR_NO_MEMORY);
-
-    uint32_t nocLen = 0;
-    ChipLogProgress(Controller, "Generating operational certificate for device " ChipLogFormatX64, ChipLogValueX64(deviceId));
-    ReturnErrorOnFailure(mOperationalCredentialsDelegate->GenerateNodeOperationalCertificate(
-        PeerId().SetNodeId(deviceId), CSR, 1, noc.Get(), kMaxCHIPDERCertLength, nocLen));
-
-    ReturnErrorCodeIf(nocLen == 0, CHIP_ERROR_CERT_NOT_FOUND);
-
     chip::Platform::ScopedMemoryBuffer<uint8_t> ica;
     ReturnErrorCodeIf(!ica.Alloc(kMaxCHIPDERCertLength), CHIP_ERROR_NO_MEMORY);
 
-    uint32_t icaLen = 0;
+    MutableByteSpan icaCertSpan(ica.Get(), kMaxCHIPDERCertLength);
+
     ChipLogProgress(Controller, "Getting intermediate CA certificate from the issuer");
-    CHIP_ERROR err = mOperationalCredentialsDelegate->GetIntermediateCACertificate(0, ica.Get(), kMaxCHIPDERCertLength, icaLen);
+    CHIP_ERROR err = mOperationalCredentialsDelegate->GetIntermediateCACertificate(0, icaCertSpan);
     ChipLogProgress(Controller, "GetIntermediateCACertificate returned %" PRId32, err);
     if (err == CHIP_ERROR_INTERMEDIATE_CA_NOT_REQUIRED)
     {
         // This implies that the commissioner application uses root CA to sign the operational
         // certificates, and an intermediate CA is not needed. It's not an error condition, so
         // let's just send operational certificate and root CA certificate to the device.
-        icaLen = 0;
+        icaCertSpan.reduce_size(0);
         ChipLogProgress(Controller, "Intermediate CA is not needed");
     }
     else if (err != CHIP_NO_ERROR)
     {
         return err;
     }
 
-    ReturnErrorOnFailure(ConvertX509CertsToChipCertArray(ByteSpan(noc.Get(), nocLen), ByteSpan(ica.Get(), icaLen), cert));
+    ReturnErrorOnFailure(ConvertX509CertsToChipCertArray(noc, icaCertSpan, cert));
 
     return CHIP_NO_ERROR;
 }
 
+void DeviceController::OnLocalNOCGenerated(void * context, const ByteSpan & noc, const PeerId & deviceId)
```

Comment:
`deviceId` argument is not used in method. Suggest removing.

Reply:
`deviceId/nodeId` will be useful when we remove it from `PairDevice` API. The commissioner will need this to maintain the state of local/proxy device object list.

In this case, it'll be used to find the commissioner's node ID.

Reply:
removed nodeId from the callback.